## Fuzzy Matching and Fuzzy Pandas
      
[Max Harlow](https://twitter.com/maxharlow), a journalist at the Financial Times, wrote this library `csvmatch`, and he's been adding new algorithms to facilitate fuzzy matching across datasets. He's used it for a bunch of stories, including:
- https://www.theguardian.com/uk-news/2014/jul/09/offshore-tax-dealings-celebrities-sportsmen-leaked-jersey-files
- https://www.theguardian.com/politics/2014/jul/08/offshore-secrets-wealthy-political-donors

Similar techniques have also been used in other stories like:
- https://www.globalwitness.org/en/campaigns/oil-gas-and-mining/myanmarjade
- https://www.irinnews.org/investigation/2016/09/02/exclusive-un-paying-blacklisted-diamond-company-central-african-republic

But, wait, first: 

### What is Fuzzy Matching? 

Automating the look-up for names in documents is [inherently imprecise](https://www.elastic.co/blog/found-fuzzy-search). The computer can't _know_ that different representations of the same _thing_ refer to the same _thing_. For example: 
- _Apple Inc._; _Apple Computer Company_; _Apple Computer, Inc._; and _Apple_ all refer to the fruit company. 
- _Samuel Langhorne Clemens_, _Samuel L. Clemens_, _Samuel Clemens_ and Mark Twain all refer to the same person. 
- _Robert Ford_, _Rob Ford_, and _Robert Frod_ refer to the same person **probably**. 

When you're working with unstructured data, you can't take anything for granted. Least of all, you can't assume that:
- documents will have correct spellings
- first, last, and middle names will exist in all documents
- the abbreviated/shortened names of people won't make an appearance (e.g. Jon instead of Jonathan, Tom instead of Thomas, Phil instead of Philip, etc.) 

So, when you're living in an uncertain world, you try to make things slightly more _certain_ with **Fuzzy Matching**. You might not hit 100 percent, but at least you'll hit more than what you would without fuzzy matching. 

There are multiple algorithms that try to minimise the uncertainty/enable fuzzy matching. The library we are going to be look at today incorporates a bunch of these, instead of just doing one thing. 

This notebook's predominantly based on an [awesome NICAR2019 presentation](https://docs.google.com/presentation/d/1djKgqFbkYDM8fdczFhnEJLwapzmt4RLuEjXkJZpKves/) where Max Harlow (the aforementioned news app developer at the Financial Times) demonstrated [csvmatch](https://github.com/maxharlow/csvmatch). And, then, Soma basically created a library to make it with Pandas. 

Worth remembering that there are no shortcuts in life, and few panaceas. Depending on the project you're working on, you might be more inclined to use one algorithm or the other. Or, you know, try a few of them and see what happens. And, also, remember: all computational tools you use need to hand-in-hand with traditional reporting. People share names, there's more than one John Smith, etc. 


In [1]:
# Make sure you `pip install fuzzy_pandas` first. 

import pandas as pd
import fuzzy_pandas as fpd

### A Toy Example

We'll be working with two toy datasets first, just to get going and get an idea as to what's possible. The names of the files are not terribly imaginative: `data1.csv` and `data2.csv`. And, they both contain structured data: names, code names and locations of characters from John le Carré's spy thriller: Tinker Tailor Soldier Spy. 

Right, let's have a look. 


In [2]:
df1 = pd.read_csv("data/data1.csv")
df2 = pd.read_csv("data/data2.csv")

In [3]:
df1.sort_values("name")

name  location   codename
5     Bill Haydon    London     Tailor
8    Connie Sachs    Oxford       none
0   George Smiley    London  Beggerman
7    Jim Prideaux  Slovakia       none
6    Oliver Lacon    London       none
1  Percy Alleline    London     Tinker
4   Peter Guillam   Brixton       none
2       Roy Bland    London    Soldier
3  Toby Esterhase    Vienna    Poorman

In [4]:
df2.sort_values("Person Name")

Person Name Location
8           Claus Kretzschmar  Hamburg
2               George SMILEY   London
4                  Konny Saks   Oxford
0  Maria Andreyevna Ostrakova   Russia
1                Otto Leipzig  Estonia
3               Peter Guillam  Brixton
6                 Sam Collins  Vietnam
5                Saul Enderby   London
7              Tony Esterhase   Vienna

### Exact matches

We start with doing "exact matches", i.e. both tables should have the exact same name. Capitalisation matters, accents matter. With this function, for example:
- John le Carre will not match with John le Carré
- George SMILEY will not match with George Smiley

Based on what you see in the data frames above, how many matches do you expect? 

In [5]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name')

name location codename    Person Name Location
0  Peter Guillam  Brixton     none  Peter Guillam  Brixton

Right, so, we only find one match as expected. But, are there any other matches that a _smarter_ algorithm could find? Let's try something called **Levenshtein**, a nifty simple algorithm that's pretty common. It's the basis for a bunch of spellcheck algorithms, amongst other things, and the way it works is it checks the number of characters that are different between two inputs, and if the _distance_ is small enough, it assumes the two words are the same. 

For example, in the above two data frames, you have Toby Esterhase and Tony Esterhase, which means the Levenshtein distance is 1 (The 'b' v. 'n' in To(b,n)y.). 

In [6]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name', method='levenshtein')

name location   codename     Person Name Location
0   George Smiley   London  Beggerman   George SMILEY   London
1  Toby Esterhase   Vienna    Poorman  Tony Esterhase   Vienna
2   Peter Guillam  Brixton       none   Peter Guillam  Brixton

The other thing you'll notice above is that, by default, the _Levenshtein_ algorithm doesn't care about case. 

However, are we still missing potential matches? 

When we work with any algorithms, we need a confidence threshold that we decide on. By default, the `csvmatch` algorithm has a `threshold` of 0.6, i.e. only if the algorithm returns a match score greater than or equal to 0.6 will it return a match. 

The score, in this case, is calculated using the below formula: 

> `1 - (distance/maximum(value1, value2))`

We can be slightly more conservative with the threshold, and we get a Brand New Result in our output. 

In [7]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name', method='levenshtein', threshold=0.5)

name location   codename     Person Name Location
0   George Smiley   London  Beggerman   George SMILEY   London
1  Toby Esterhase   Vienna    Poorman  Tony Esterhase   Vienna
2   Peter Guillam  Brixton       none   Peter Guillam  Brixton
3    Connie Sachs   Oxford       none      Konny Saks   Oxford

This is _cool_. By changing the threshold, we found another match based on what the pronunciation of the two names are: Connie and Konny. **But, what could be cooler?**

In [8]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name', method='metaphone')

name location   codename    Person Name Location
0  George Smiley   London  Beggerman  George SMILEY   London
1  Peter Guillam  Brixton       none  Peter Guillam  Brixton
2   Connie Sachs   Oxford       none     Konny Saks   Oxford

The **metaphone** algorithm does phonetic matching, and gives you results based on that. 

Note: In theory, the documentation says that you can combine a couple of these algorithms if you're so inclined. But, it looks like when you combine two algorithms, it doesn't _quite_ work. ¯\_(ツ)_/¯

In [9]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name', method=['levenshtein', 
                                                                          'metaphone'])

## swap the methods around and then look at the results, too.

name location   codename     Person Name Location
0   George Smiley   London  Beggerman   George SMILEY   London
1  Toby Esterhase   Vienna    Poorman  Tony Esterhase   Vienna
2   Peter Guillam  Brixton       none   Peter Guillam  Brixton

What do you think is happening here? 

This is important—you're often going to be using tools built by other folks, but where there's code, there are bugs. You should make sure that you play with the tool a bit to make sure it's doing _exactly_ what you think it's doing. And, if it's not, you know where it falls short. 

### Less Fictional Datasets

We are going to be using the same datasets Max Harlow used for this exercise. As he explains in his presentation [here](https://docs.google.com/presentation/d/1djKgqFbkYDM8fdczFhnEJLwapzmt4RLuEjXkJZpKves/edit#slide=id.g3512a0ce6b_1_22), there are a bunch of files: 
- a list of world billionaires published by Bloomberg
- a similar list published by Forbes
- a list also published by Forbes that only includes Chinese individuals
- a list published by the CIA of chiefs of state and cabinet members of foreign governments
- a list of all the people that attended the World Economic Forum conference in Davos this year
- a list of all the people and companies that have been sanctioned by the United Nations


In [10]:
## Read in the two billionaire lists (Forbes + Bloomberg)

df1 = pd.read_csv("data/forbes-billionaires.csv")
df2 = pd.read_csv("data/bloomberg-billionaires.csv")

Can you find out how many billionaires appear in both lists (exact matching)? 

In [11]:
df1.sample(30)

name     lastName                   uri  \
2306              Wang Yong         Wang             wang-yong   
194           Richard Sands        Sands         richard-sands   
1643   Gustav Magnar Witzoe       Witzoe  gustav-magnar-witzoe   
137             Koo Bon-sik          Koo           koo-bon-sik   
1207             Yuan Yafei         Yuan            yuan-yafei   
649         Leandro Rizzuto      Rizzuto       leandro-rizzuto   
1990               Huang Yi        Huang              huang-yi   
901           Riley Bechtel      Bechtel         riley-bechtel   
1070           Najib Mikati       Mikati          najib-mikati   
2162       Andrei Rappoport    Rappoport      andrei-rappoport   
2292        O. Bruton Smith        Smith        o-bruton-smith   
2060           Zhang Xiugen        Zhang          zhang-xiugen   
347         Abigail Johnson      Johnson       abigail-johnson   
1813           Marina Prada        Prada          marina-prada   
1541        Daniel Roullier     Roullier       daniel-roullier   
1101             Bill Gross        Gross            bill-gross   
1208           Zhou Yaoting         Zhou          zhou-yaoting   
1306         Dermot Desmond      Desmond        dermot-desmond   
1931        Kenneth Tuchman      Tuchman       kenneth-tuchman   
340   Thomas & Raymond Kwok         Kwok   thomas-raymond-kwok   
1096           Kenneth Feld         Feld          kenneth-feld   
1061           Doris Fisher       Fisher          doris-fisher   
886    Sudhir & Samir Mehta        Mehta    sudhir-samir-mehta   
1195          Michael Rubin        Rubin         michael-rubin   
1486        J. Joe Ricketts     Ricketts        j-joe-ricketts   
1518      Fritz Draexlmaier  Draexlmaier     fritz-draexlmaier   
1175            Gerald Ford         Ford           gerald-ford   
2288             Evan Sharp        Sharp            evan-sharp   
2024            Chris Sacca        Sacca           chris-sacca   
2259          Lee Sang-Hyuk          Lee         lee-sang-hyuk   

                 imageUri  worthChange                          source  \
2306            wang-yong          NaN                 food sweeteners   
194                no-pic      -13.513                 Food & Beverage   
1643               no-pic        9.969                    fish farming   
137           koo-bon-sik       22.025  electronics/household products   
1207           yuan-yafei          NaN             retail, investments   
649                no-pic          NaN               consumer products   
1990               no-pic       95.983               auto distribution   
901         riley-bechtel        0.000       engineering, construction   
1070         najib-mikati        0.524                         telecom   
2162     andrei-rappoport        0.000                     investments   
2292               no-pic          NaN               Nascar racetracks   
2060               no-pic        0.000                     automobiles   
347       abigail-johnson        0.000                money management   
1813         marina-prada       -3.918                    luxury goods   
1541               no-pic          NaN                     animal feed   
1101           bill-gross        0.000                     investments   
1208               no-pic       10.448            apparel, real estate   
1306       dermot-desmond      -35.869                         finance   
1931               no-pic       13.844                     outsourcing   
340   thomas-raymond-kwok      353.471                     real estate   
1096         kenneth-feld        0.000      circus, live entertainment   
1061         doris-fisher       -4.151                             Gap   
886    sudhir-samir-mehta       40.464                     diversified   
1195        michael-rubin        0.000                   online retail   
1486       j-joe-ricketts        5.426                   TD Ameritrade   
1518    fritz-draexlmaier        0.000              

In [12]:
df2.sample(30)

Rank                        Name Total_net_worth             Country  \
441   442                   Bill Koch          $4.07B       United States   
43     44              Rupert Murdoch          $18.4B       United States   
384   385                 Chris Reyes          $4.47B       United States   
57     58                 Lee Kun Hee          $15.9B  Korea, Republic of   
331   332               David Thomson          $4.84B              Canada   
325   326              Quek Leng Chan          $4.91B            Malaysia   
160   161             Pam Mars-Wright          $8.37B       United States   
271   272                Henry Kravis          $5.68B       United States   
114   115                  Uday Kotak          $11.1B               India   
33     34          Jorge Paulo Lemann          $21.3B              Brazil   
116   117             Reinhold Wuerth          $11.0B             Germany   
118   119                   Peter Woo          $10.8B           Hong Kong   
336   337                Kenneth Dart          $4.77B      Cayman Islands   
396   397               Denise Coates          $4.36B      United Kingdom   
397   398              Guo Guangchang          $4.36B               China   
391   392  Magdalena Martullo-Blocher          $4.42B         Switzerland   
5       6                 Carlos Slim          $59.0B              Mexico   
90     91          Andrey Melnichenko          $12.8B  Russian Federation   
318   319               Leonard Stern          $4.97B       United States   
265   266              John P Grayken          $5.80B             Ireland   
289   290               Arthur Irving          $5.41B              Canada   
426   427                   David Sun          $4.15B       United States   
474   475                 Mary Malone          $3.84B       United States   
480   481            Ralph Dommermuth          $3.80B             Germany   
231   232                Seo Jung-Jin          $6.41B  Korea, Republic of   
439   440               Jeremy Jacobs          $4.09B       United States   
242   243         Mike Cannon-Brookes          $6.27B           Australia   
126   127                   Li Shu Fu          $10.4B               China   
403   404                  Vikram Lal          $4.34B               India   
487   488                Ajay Piramal          $3.75B               India   

            Industry  
441      Diversified  
43   Media & Telecom  
384         Consumer  
57       Diversified  
331  Media & Telecom  
325      Diversified  
160  Food & Beverage  
271          Finance  
114          Finance  
33   Food & Beverage  
116       Industrial  
118      Real Estate  
336          Finance  
396    Entertainment  
397          Finance  
391       Industrial  
5        Diversified  
90        Industrial  
318      Real Estate  
265          Finance  
289           Energy  
426       Technology  
474  Food & Beverage  
480       Technology  
231      Health Care  
439  Food & Beverage  
242       Technology  
126       Industrial  
403       Industrial  
487          Finance

In [13]:
results = fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Name')

print("Found", results.shape)
results.head(5)

Found (354, 19)


name   lastName               uri imageUri  worthChange  \
0    Alexander Otto       Otto    alexander-otto   no-pic        2.120   
1     Ben Ashkenazy  Ashkenazy     ben-ashkenazy   no-pic        0.000   
2  Giovanni Ferrero    Ferrero  giovanni-ferrero   no-pic        0.000   
3       Henry Cheng      Cheng     henry-cheng-1   no-pic        3.542   
4      Henry Laufer     Laufer      henry-laufer   no-pic        0.000   

                source                 industry gender        country  \
0          real estate              Real Estate      M        Germany   
1          real estate              Real Estate      M  United States   
2  Nutella, chocolates        Food and Beverage      M          Italy   
3             property              Diversified      M      Hong Kong   
4          hedge funds  Finance and Investments      M  United States   

       timestamp  realTimeWorth  realTimeRank  realTimePosition  \
0  1547575201867      10821.927         126.0             126.0   
1  1547574901333       4000.000         499.0             499.0   
2  1547575201866      22673.165          38.0              38.0   
3  1547574901334       1334.282        1630.0            1630.0   
4  1547574901333       2000.000        1141.0            1142.0   

                                         squareImage  Rank              Name  \
0  //specials-images.forbesimg.com/imageserve/5a7...   323    Alexander Otto   
1  //specials-images.forbesimg.com/imageserve/59e...   447     Ben Ashkenazy   
2  //specials-images.forbesimg.com/imageserve/5b1...    33  Giovanni Ferrero   
3  //specials-images.forbesimg.com/imageserve/5a7...    79       Henry Cheng   
4                                                NaN   463      Henry Laufer   

  Total_net_worth        Country         Industry  
0          $4.94B        Germany      Real Estate  
1          $4.05B  United States      Real Estate  
2          $22.6B          Italy  Food & Beverage  
3          $14.1B      Hong Kong           Retail  
4          $3.95B  United States          Finance

Now, can you find the ones where the ranks aren't the same across the two datasets? What about the ones that are the same?

In [14]:
results = fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Name', 
                          keep_left=["realTimeRank", "name"], 
                          keep_right=["Rank"])
results[results.realTimeRank == results.Rank]

realTimeRank              name  Rank
13            2.0        Bill Gates     2
14            3.0    Warren Buffett     3
15            1.0        Jeff Bezos     1
21            4.0   Bernard Arnault     4
23           10.0       Sergey Brin    10
27           23.0   Sheldon Adelson    23
37           17.0     Mukesh Ambani    17
47           54.0     Stefan Quandt    54
88          135.0     Aliko Dangote   135
131         177.0  Cyrus Poonawalla   177
149         139.0      Gordon Moore   139
204         235.0      Ralph Lauren   235

### Fuzzy matching with non-fictional data

In the above couple of cells, we've conducted "exact matching", i.e. the equivalent of you running a `Cmd+F`/`Ctrl+F` on your text editor. But, this is _almost_ worse as it's case sensitive, i.e. "Tom" and "tom" are treated differently. 

We've gone through some of this already, but what are the things we can ignore when it comes to name-matching? Harlow, in his presentation, identified:
- case
- title (Mr., Mrs., etc.)
- non-latin characters (é, å, ß, etc.)
- the order of the names
- non-alphanumerics (e.g. hyphenated names)

Now, you don't _have to_ ignore _anything_, but sometimes, it might make your life far easier. Other times, you'll end up with false positives and whatnot. 

The library `csvmatch`—and by extension `fuzzy_pandas`—support a bunch of the above parameters, which you can just pass in to the function. Passing in a bunch of these parameters would allow you to go from `Orbán, Viktor` to `Viktor Orban`, which is quite useful. (Again, the example's from Harlow's slides)

For this bit, we'll move on to two of the other datasets: `cia-world-leaders.csv` and `davos-attendees-2019.csv`. As always, read in the data and figure out which columns the exact match should run on. 

In [15]:
cia_world_leaders = pd.read_csv('data/cia-world-leaders.csv')
davos_attendees = pd.read_csv('data/davos-attendees-2019.csv')
print(f"Our CIA World Leaders df has these columns: {cia_world_leaders.columns} \
      \n The Davos attendees have these: {davos_attendees.columns}")

Our CIA World Leaders df has these columns: Index(['country', 'role', 'name'], dtype='object')       
 The Davos attendees have these: Index(['full_name', 'position_short_name', 'org_name', 'org_country'], dtype='object')


In [16]:
cia_world_leaders.sort_values('name').head(20)

country                                               role  \
384             Bangladesh                            Min. of Foreign Affairs   
499                 Belize                             Governor, Central Bank   
4721               Somalia              Min. of Public Works & Reconstruction   
4451          Saudi Arabia                                   Min. of Interior   
2625                Jordan                                               King   
4193                 Qatar                                        Deputy Amir   
4213                 Qatar                                         Prime Min.   
4205                 Qatar                                   Min. of Interior   
4194                 Qatar                       Governor, Qatar Central Bank   
5482  United Arab Emirates  Min. of Foreign Affairs and International Coop...   
5487  United Arab Emirates                 Min. of Infrastructure Development   
1809              Ethiopia                                         Prime Min.   
1732               Eritrea                               Min. of Public Works   
4891                 Sudan                      Min. of Guidance & Endowments   
1799              Ethiopia                                    Min. of Revenue   
5139              Thailand                                      Min. of Labor   
394             Bangladesh                                   Min. of Planning   
366                Bahrain      Min. of the Royal Court for Follow-Up Affairs   
4199                 Qatar                            Min. of Economy & Trade   
353                Bahrain                                    Min. of Finance   

                                                   name  
384                                   A. H. Mahmood ALI  
499                                        A. Joy GRANT  
4721                     ABAS Abdullahi Sheikh "Siraji"  
4451  ABD AL-AZIZ bin Saud bin Nayif bin Abd al-Aziz...  
2625                                        ABDALLAH II  
4193                        ABDALLAH bin Hamad Al Thani  
4213            ABDALLAH bin Nasir bin Khalifa Al Thani  
4205            ABDALLAH bin Nasir bin Khalifa Al Thani  
4194                         ABDALLAH bin Saud Al Thani  
5482                     ABDALLAH bin Zayid Al Nuhayyan  
5487             ABDULLAH bin Muhammad Belhaf al-Nuaymi  
1809                                         ABIY Ahmed  
1732                                      ABRAHA Asfaha  
4891                            ABU-BAKR Uthman Ibrahim  
1799                                      ADANECH Abebe  
5139                      ADUN Saengsingkaew, Pol. Gen.  
394                                   AHM Mostafa KAMAL  
366                    AHMAD BIN ATIYATALLAH Al Khalifa  
4199              AHMAD bin Jasim bin Muhammad Al Thani  
353   AHMAD bin Muhammad bin Hamad bin Abdallah Al K...

In [17]:
davos_attendees.sort_values('full_name').head(20)

full_name  \
1864                 Aaron Karczmer   
2532               Aaron Motsoaledi   
422              Aarthi Subramanian   
2074            Abdelkader Messahel   
897            Abdulaziz Al Judaimi   
888            Abdulaziz Al Subeaei   
35            Abdulaziz Al-Helaissi   
903             Abdulaziz Al-Jarbou   
2641               Abdulla Al Basti   
809              Abdulla Al Khalifa   
1228             Abdulla Bin Desmal   
804   Abdulla Bin Mohammed Al Thani   
2638               Abdulla Bin Touq   
2660          Abdulla Nasser Lootah   
2071              Abdullah Abdullah   
895          Abdulmajeed Al Mobarak   
663                Abdulsamad Rabiu   
2428                     Abeer Odeh   
2297               Abha Joshi-Ghani   
695                       Abid Butt   

                                    position_short_name  \
1864  Executive Vice-President; Chief Risk, Complian...   
2532                 Minister of Health of South Africa   
422              Executive Director, Board of Directors   
2074             Minister of Foreign Affairs of Algeria   
897                   Senior Vice-President, Downstream   
888                                            Chairman   
35                  Group Chief Executive Officer (GIB)   
903                               Chairman of the Board   
2641  General Secretary, Executive Council of Dubai,...   
809                             Chief Executive Officer   
1228                    Director, Commercial Operations   
804                                            Chairman   
2638  Secretary-General, Cabinet of the United Arab ...   
2660  Director-General, Federal Competitiveness and ...   
2071  Chief Executive of the Islamic Republic of Afg...   
895                             Chief Executive Officer   
663                Chairman and Chief Executive Officer   
2428  Minister of National Economy of the Palestinia...   
2297  Senior Adviser, Infrastructure, Public- Privat...   
695                             Chief Executive Officer   

                                               org_name  \
1864                                             PayPal   
2532                 Ministry of Health of South Africa   
422                           Tata Consultancy Services   
2074             Ministry of Foreign Affairs of Algeria   
897                                        Saudi Aramco   
888                     Jabal Omar Development Company    
35                          Gulf International Bank BSC   
903                 Saudi Basic Industries Corporation    
2641                         Executive Council of Dubai   
809                        Qatar National Bank Q.P.S.C.   
1228                                   DP World Limited   
804                                             Ooredoo   
2638                Cabinet of the United Arab Emirates   
2660  The Federal Competitiveness and Statistics Aut...   
2071                             Afghanistan Government   
895                                          Riyad Bank   
663                                           BUA Group   
2428  Ministry of National Economy of the Palestinia...   
2297                                     The World Bank   
695                         e2e Supply Chain Management   

                  org_country  
1864                      USA  
2532             South Africa  
422                     India  
2074                  Algeria  
897              Saudi Arabia  
888              Saudi Arabia  
35                    Bahrain  
903              Saudi Arabia  
2641     United Arab Emirates  
809                     Qatar  
1228     United Arab Emirates  
804                     Qatar  
2638     United Arab Emirates  
2660     United Arab Emirates  
2071              Afghanistan  
895              Saudi Arabia  
663                   Nigeria  
2428  Palestinian Territories  
2297                      INT  
695                  Pakistan

In [18]:
# Let's start by seeing what an exact match would look like. 
results = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', right_on='full_name')

Apparently, there are no matches. What do you reckon? What _should_ the overlap between CIA's list of world leaders and Davos attendees be? 

Let's try some _fuzzy matching_. 

In [19]:
results = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', 
                          right_on='full_name', 
                          ignore_case=True)

In [20]:
print(results.shape)
results

(119, 7)


country                                               role  \
0           Algeria  Min. of Foreign Affairs & International Cooper...   
1         Argentina                          Min. of Production & Work   
2         Argentina                                Pres., Central Bank   
3           Armenia                                         Prime Min.   
4         Australia                          Min. for Defense Industry   
5         Australia              Min. for Trade, Investment, & Tourism   
6           Austria                                         Chancellor   
7           Austria    Min. for Europe, Integration, & Foreign Affairs   
8        Azerbaijan                                              Pres.   
9           Belgium                                    Dep. Prime Min.   
10          Belgium                                    Dep. Prime Min.   
11          Belgium  Min. of Development Cooperation, Digital Agend...   
12          Belgium    Min. of Employment, Economy, & Consumer Affairs   
13            Burma    Min. of Investment & Foreign Economic Relations   
14         Cambodia              Min. of Public Works & Transportation   
15           Canada                                    Min. of Finance   
16           Canada                            Min. of Foreign Affairs   
17            China                                    Chief Executive   
18            China          Min. of Industry & Information Technology   
19            China                                   Sec. for Finance   
20       Costa Rica                                              Pres.   
21          Denmark                                Min. for Employment   
22          Denmark                                   Min. for Finance   
23         Ethiopia             Min. of Finance & Economic Cooperation   
24         Ethiopia                                         Prime Min.   
25             Fiji       Permanent Representative to the UN, New York   
26          Finland                                    Dep. Prime Min.   
27          Finland                           Min. for Foreign Affairs   
28           France                  Min. for Europe & Foreign Affairs   
29           France                          Min. of Economy & Finance   
..              ...                                                ...   
89        Singapore  Coordinating Min. for Economic & Social Polici...   
90        Singapore                                    Dep. Prime Min.   
91        Singapore                                   Min. for Finance   
92         Slovakia                                         Prime Min.   
93     South Africa                       Min. of Economic Development   
94     South Africa                         Min. of Public Enterprises   
95            Spain              Min. of Industry, Commerce, & Tourism   
96      Switzerland  Chief, Federal Dept. of Defense, Civil Protect...   
97      Switzerland                    Chief, Federal Dept. of Finance   
98      Switzerland            Chief, Federal Dept. of Foreign Affairs   
99      Switzerland               Chief, Federal Dept. of Home Affairs   
100     Switzerland                                 Federal Chancellor   
101     Switzerland                         Pres., Swiss Confederation   
102     Switzerland                                         Vice Pres.   
103         Tunisia                                         Prime Min.   
104          Turkey                                      Min. of Trade   
105          Turkey                         Min. of Treasury & Finance   
106          Uganda                                              Pres.   
107         Ukraine                            Min. of Foreign Affairs   
108         Ukraine                                              Pres.   
109  United Kingdom                          Governor, Bank of England   
110  United Kingdom  Sec. of State for Business, Energy, & Industri...   
111  United King

OK, we have more matches, but this is also pretty boring. There's nothing super-smart about ignoring cases to get matches. Your word processors have been doing that for _decades_. 

But, now, let's start adding some of our other parameters discussed above, and see what happens

In [21]:
results = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', 
                          right_on='full_name', 
                          ignore_case=True, 
                          ignore_nonalpha=True,
                          ignore_nonlatin=True,
                          ignore_order_words=True,
                          ignore_titles=True,
                         )

In [22]:
print(results.shape)
results

(138, 7)


country  \
0                   Algeria   
1                 Argentina   
2                 Argentina   
3                 Argentina   
4                   Armenia   
5                 Australia   
6                 Australia   
7                   Austria   
8                   Austria   
9                Azerbaijan   
10                  Belgium   
11                  Belgium   
12                  Belgium   
13                  Belgium   
14   Bosnia and Herzegovina   
15                    Burma   
16                 Cambodia   
17                   Canada   
18                   Canada   
19                    China   
20                    China   
21                    China   
22               Costa Rica   
23            Cote d'Ivoire   
24                  Croatia   
25                     Cuba   
26                  Czechia   
27                  Denmark   
28                  Denmark   
29                  Denmark   
..                      ...   
108                Slovakia   
109            South Africa   
110            South Africa   
111                   Spain   
112                   Spain   
113             Switzerland   
114             Switzerland   
115             Switzerland   
116             Switzerland   
117             Switzerland   
118             Switzerland   
119             Switzerland   
120                 Tunisia   
121                  Turkey   
122                  Turkey   
123                  Turkey   
124                  Turkey   
125                  Uganda   
126                 Ukraine   
127                 Ukraine   
128          United Kingdom   
129          United Kingdom   
130          United Kingdom   
131          United Kingdom   
132          United Kingdom   
133          United Kingdom   
134                 Vietnam   
135                 Vietnam   
136                 Vietnam   
137                Zimbabwe   

                                                  role  \
0    Min. of Foreign Affairs & International Cooper...   
1                            Min. of Production & Work   
2                           Min. of Treasury & Finance   
3                                  Pres., Central Bank   
4                                           Prime Min.   
5                            Min. for Defense Industry   
6                Min. for Trade, Investment, & Tourism   
7                                           Chancellor   
8      Min. for Europe, Integration, & Foreign Affairs   
9                                                Pres.   
10                                     Dep. Prime Min.   
11                                     Dep. Prime Min.   
12   Min. of Development Cooperation, Digital Agend...   
13     Min. of Employment, Economy, & Consumer Affairs   
14                         Chmn., Council of Ministers   
15     Min. of Investment & Foreign Economic Relations   
16               Min. of Public Works & Transportation   
17                                     Min. of Finance   
18                             Min. of Foreign Affairs   
19                                     Chief Executive   
20           Min. of Industry & Information Technology   
21                                    Sec. for Finance   
22                                               Pres.   
23                           Min. of Economy & Finance   
24                                          Prime Min.   
25                  Min. of Foreign Trade & Investment   
26                                          Prime Min.   
27                                 Min. for Employment   
28                                    Min. for Finance   
29                                          Prime Min.   
..                                                 ...   
108                                         Prime Min.   
109                       Min. of Economic Development   
110                         Min. of Public Enterprises   
111                 Min. of Economy & Business Affairs   
112      

Right, 19 more results. Baby steps, but at least steps in the right direction. Now, let's start using the more _intelligent_ algorithms in place—this one named after a Russian mathematician: Levenshtein. 

All *Levenshtein* does is look at how many characters are different between two inputs? For example:

In [23]:
from jellyfish._jellyfish import damerau_levenshtein_distance

damerau_levenshtein_distance("Évry", "Every")

2

Let's quickly use the algorithm directly to see the output. The above cell imports something called `jellyfish`, which is another package that `csvmatch` uses. Typically, you wouldn't call the function directly (you could if you wanted to), but this is just to give you an idea of how the algorithm works. 

Right, now let's use this with our above data, and see if we have better luck. 

Remember: the threshold specified by us is 0.6, and it's calculated by: `1-(distance/max(value1, value2))`. In the case of `Évry` and `Every` above, our calculation would be:

`1 - (2/5)` = `3/5` = `0.6`

So, in this case, the two would lead to a fuzzy match.

In [24]:
results = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', 
                          right_on='full_name', 
                          ignore_case=True, 
                          ignore_nonalpha=True,
                          ignore_nonlatin=True,
                          ignore_order_words=True,
                          ignore_titles=True,
                          method='levenshtein'
                         )

In [25]:
results.shape

(1449, 7)

**WAIT, WHAT?!** Have we just gone from 138 matches to 1952? Is that overtly optimistic? 

In [26]:
results.sample(50)

country  \
1137                Portugal   
672                   Jordan   
770                  Lebanon   
1126             Philippines   
1344                  Turkey   
482                    Gabon   
632                    Italy   
480                    Gabon   
454                     Fiji   
94                 Australia   
1400          United Kingdom   
383                   Cyprus   
1375          United Kingdom   
95                 Australia   
61                 Australia   
640                  Jamaica   
1000             New Zealand   
834                   Malawi   
1103                Paraguay   
1300                   Syria   
826               Madagascar   
1212               Singapore   
689                    Kenya   
1034             New Zealand   
1388          United Kingdom   
83                 Australia   
953                    Nauru   
164                  Belarus   
725             Korea, North   
1109                    Peru   
685                    Kenya   
43                 Argentina   
426                  Ecuador   
986              New Zealand   
684                    Kenya   
798            Liechtenstein   
1101        Papua New Guinea   
1415               Venezuela   
987              New Zealand   
863                     Mali   
708             Korea, North   
855                 Maldives   
518                  Grenada   
503                  Germany   
271                   Canada   
499                  Germany   
567                    India   
228   Bosnia and Herzegovina   
430                  Ecuador   
1092        Papua New Guinea   

                                                   role  \
1137                         Governor, Bank of Portugal   
672                                          Prime Min.   
770                          Min. of Telecommunications   
1126                                Sec. of Agriculture   
1344                                      Min. of Trade   
482                                       Min. of Mines   
632                   Min. of Cultural Assets & Tourism   
480                                       Min. of Mines   
454   Min. for Industry, Trade, Tourism, Local Govt....   
94                              Min. for Human Services   
1400                            Sec. of State for Wales   
383                                    Min. of Interior   
1375                               Min. for Immigration   
95                              Min. for Human Services   
61                                        Attorney Gen.   
640                                     Min. of Defense   
1000                                    Min. for Racing   
834                                     Min. of Defense   
1103                                    Min. of Justice   
1300                                    Dep. Prime Min.   
826              Min. of Public Works & Infrastructures   
1212               Min. for Communication & Information   
689                                          Dep. Pres.   
1034                    Min. of Trade and Export Growth   
1388  Sec. of State for Environment, Food, & Rural A...   
83                      Min. for Digital Transformation   
953          Min. for Finance & Sustainable Development   
164                                 Min. of Information   
725                            Min. of Urban Management   
1109                    Min. of Foreign Trade & Tourism   
685                  Cabinet Sec. for Sports & Heritage   
43                  Min. of Health & Social Development   
426                                     Min. of Finance   
986                                     Min. for Courts   
684    Cabinet Sec. for Industry, Trade, & Cooperatives   
798        Permanent Representative to the UN, New York   
1101                    Min. for Works & Implementation   
1415                                     Min. of Health   
987                      Min. for Crown/Maori Relations   
863            Min. of Solid

Why, yes. Yes, it is. This is why you _always_ confirm what an algorithm does. Right, maybe let's bump up our threshold and see what happens. 

In [27]:
results = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', 
                          right_on='full_name', 
                          ignore_case=True, 
                          ignore_nonalpha=True,
                          ignore_nonlatin=True,
                          ignore_order_words=True,
                          ignore_titles=True,
                          method='levenshtein', 
                          threshold=0.8
                         )

In [28]:
results.shape

(185, 7)

OK, that seems more reasonable. Let's sanity check.

In [29]:
results.sample(50)

country                                               role  \
3           Argentina                         Min. of Treasury & Finance   
65          Indonesia     Min. of Communication & Information Technology   
168           Ukraine                                              Pres.   
155       Switzerland               Chief, Federal Dept. of Home Affairs   
43              Egypt                                         Prime Min.   
92       Korea, South                          Min. of Economy & Finance   
84              Japan  Min. in Charge of Support for Women's Empowerment   
139            Serbia                            Min. of Energy & Mining   
177    United Kingdom                            Sec. of State for Wales   
122           Nigeria              Min. of Industry, Trade, & Investment   
103        Luxembourg                                      Min. of State   
20           Botswana              Min. of Investment, Trade, & Industry   
60              Ghana                                         Vice Pres.   
153       Switzerland                    Chief, Federal Dept. of Finance   
117       New Zealand                 Min. for Arts, Culture, & Heritage   
159             Syria                                      Min. of State   
50            Finland                           Min. for Foreign Affairs   
5             Armenia                                         Prime Min.   
48               Fiji       Permanent Representative to the UN, New York   
30              China          Min. of Industry & Information Technology   
179           Vietnam                           Min. of Industry & Trade   
23              Burma              Min. of the Office of the Union Govt.   
93       Korea, South  Min. of Science & ICT (Information & Communica...   
32         Costa Rica                                              Pres.   
86              Japan                Min. of State for Regulatory Reform   
62              India                          Min. of Corporate Affairs   
162            Turkey                             Governor, Central Bank   
178           Vietnam                                    Dep. Prime Min.   
6           Australia                                      Attorney Gen.   
118       New Zealand          Min. for National Security & Intelligence   
53             France                          Min. of Economy & Finance   
136            Rwanda                                              Pres.   
104        Luxembourg                                         Prime Min.   
29              China          Min. of Human Resources & Social Security   
124  Papua New Guinea                                      Attorney Gen.   
154       Switzerland            Chief, Federal Dept. of Foreign Affairs   
121       New Zealand                                         Prime Min.   
76             Israel                            Min. of Foreign Affairs   
59            Germany                                    Min. of Defense   
42              Egypt             Min. of Investment & Intl. Cooperation   
45           Ethiopia                                    Min. of Defense   
52             France                  Min. for Europe & Foreign Affairs   
105         Macedonia                                         Prime Min.   
147          Slovakia                                         Prime Min.   
174    United Kingdom             Sec. of State for Health & Social Care   
142         Singapore  Coordinating Min. for Economic & Social Polici...   
176    United Kingdom  Sec. of State for Intl. Trade & President of t...   
99          Lithuania                                              Pres.   
80              Italy                            Min. of Foreign Affairs   
85              Japan                  Min. of State for Gender Equality   

                        name                full_name  \
3            Nicolas DUJOVNE          Nicolás Dujovne   
65                RUDIA

Next question, one for you guys to do:

Which names from the CIA world leaders list are also on the Forbes billionaires list?

Who can get the best result?

In [30]:
forbes_df = df1

results = fpd.fuzzy_merge(cia_world_leaders, forbes_df, left_on='name', 
                          right_on='name', 
#                           ignore_case=True, 
#                           ignore_nonalpha=True,
#                           ignore_nonlatin=True,
#                           ignore_order_words=True,
#                           ignore_titles=True,
                          method='levenshtein', 
                         )

In [31]:
results.shape

(323, 17)

So in what scenarios do you reckon Levenshtein will perform badly? 

Next up: **metaphone**. 

Metaphone's great for names which sound similar, which wouldn't be caught by Levenshtein. It's especially handy when you're working with transcript data. But, it too comes with its pitfalls. 

Let's look at an example and then discuss what the possible pitfalls could be. 

Which names from the CIA world leaders list are also on the United Nations sanctions list?

In [32]:
un_sanctions = pd.read_csv("data/un-sanctions.csv")

In [33]:
results = fpd.fuzzy_merge(cia_world_leaders, un_sanctions, left_on='name', 
                          right_on='name', 
                          method='metaphone', 
                         )

In [34]:
results.shape

(18, 22)

In [35]:
## How does this compare to our other algorithms? 
results = fpd.fuzzy_merge(cia_world_leaders, un_sanctions, left_on='name', 
                          right_on='name', 
                          method='levenshtein', 
                         )

In [36]:
results.shape

(11, 22)

In [37]:
## How does this compare to our other algorithms? 
results = fpd.fuzzy_merge(cia_world_leaders, un_sanctions, left_on='name', 
                          right_on='name', 
                          ignore_case=True
                         )

In [38]:
results.shape

(6, 22)

Finally, we get to **Bilenko**. And, yes, this uses machine learning, where you train your own data. So, now you have human *smarts* being involved in the process of matching up names across documents. 

Let's look at an example: 

Which names from the CIA world leaders list are also on the Davos attendees list?

In [ ]:
results = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', 
                          right_on='full_name', 
                          method='bilenko', 
                         )


Answer questions as follows:
 y - yes
 n - no
 s - skip
 f - finished

name: Nguyen Xuan CUONG

name: Nguyen Xuan Phuc

Do these records refer to the same thing? [y/n/s/f] 

In [ ]:
results.shape